README.md
rackspace_openssh Cookbook
================
Installs and configures openssh.


Requirements
------------
### Platform
- Debian/Ubuntu
- RHEL/CentOS


Recipes
-------
### default
Selects the packages to install by package name and manages the sshd service.


Usage
-----
Ensure that the openssh packages are installed and the service is managed with `recipe[rackspace_openssh]`.


Attributes List
---------------
The attributes list is dynamically generated, and lines up with the default openssh configs.

This means anything located in [sshd_config](http://www.openbsd.org/cgi-bin/man.cgi?query=sshd_config&sektion=5) or [ssh_config](http://www.openbsd.org/cgi-bin/man.cgi?query=sshd_config&sektion=5) can be used in your node attributes.

* If the option can be entered more then once, use an _Array_, otherwise, use a _String_. If the option is host-specific use a `Hash` (please see below for more details).
* The current default attributes are opinated `ssh_config` and `sshd_config` provided by openssh.
* The namespace for `sshd_config` is `node['rackspace_openssh']['templates_cookbook']['server']`.
* Likewise, the namespace for `ssh_config` is `node['rackspace_openssh']['templates_cookbook']['client']`.
* An attribute can be an `Array`, a `Hash` or a `String`.
* If it is an `Array`, each item in the array will get it's own line in the config file.
* `Hash` attributes are meant to used with `ssh_config` namespace to create host-specific configurations. The keys of the `Hash` will be used as the `Host` entries and their associated entries as the configuration values.
* All the values in openssh are commented out in the `attributes/default.rb` file for a base starting point.


Dynamic ListenAddress
---------------------
Pass in a `Hash` of interface names, and IP address type(s) to bind sshd to. This will expand to a list of IP addresses which override the default `node['rackspace_openssh']['templates_cookbook']['server']['listen_address']` value.


Examples and Common usage
-------------------------
These can be mixed and matched in roles and attributes.  Please note, it is possible to get sshd into a state that it will not run.  If this is the case, you will need to login via an alternate method and debug sshd like normal.

####  Bind to a specific set of address (this example actually binds to all).

Not to be used with `node['rackspace_openssh']['templates_cookbook']['listen_interfaces']`.

```json
"rackspace_openssh": {
  "server": {
    "address_family": "any",
      "listen_address": [ "192.168.0.1", "::" ']
    }
  }
}
```

### Bind to the addresses tied to a set of interfaces.

```json
"rackspace_openssh": {
  "listen_interfaces": {
    "eth0": "inet",
    "eth1": "inet6"
  }
}
```

### Host-specific configurations with hashes.
You can use a `Hash` with `node['rackspace_openssh']['templates_cookbook']['client']` to configure different values for different hosts.

```json
"client": {
  "*": {
    "g_s_s_a_p_i_authentication": "yes",
    "send_env": "LANG LC_*",
    "hash_known_hosts": "yes"
  },
  "localhost": {
    "user_known_hosts_file": "/dev/null",
    "strict_host_key_checking": "no"
  },
  "127.0.0.1": {
    "user_known_hosts_file": "/dev/null",
    "strict_host_key_checking": "no"
  },
  "other*": {
    "user_known_hosts_file": "/dev/null",
    "strict_host_key_checking": "no"
  }
}
```

The keys are used as values with the `Host` entries. So, the configuration fragment shown above generates:

```text
Host *
SendEnv LANG LC_*
HashKnownHosts yes
GSSAPIAuthentication yes
Host localhost
StrictHostKeyChecking no
UserKnownHostsFile /dev/null
Host 127.0.0.1
StrictHostKeyChecking no
UserKnownHostsFile /dev/null
Host other*
StrictHostKeyChecking no
UserKnownHostsFile /dev/null
```


License & Authors
-----------------
- Author:: Adam Jacob <adam@opscode.com>
- Author:: Ted Neykov <ted.neykov@rackspace.com>
- Author:: Ryan Richard <ryan.richard@rackspace.com>

```text
Copyright:: 2008-2009, Opscode, Inc
Copyright:: 2014, Rackspace, US Inc.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
openssh Cookbook CHANGELOG
==========================
This file is used to list changes made in each version of the openssh cookbook.

v3.1.1
------
- Fix template syntax error for [Issue #4](https://github.com/rackspace-cookbooks/rackspace_openssh/issues/4)

v3.1.0
------
- Sets "PermitRootLogin" to default to "No"
- Update to latest Gemfile, Rakefile and rubocop config
- Remove logic from attributes for non-supported OS's
- Fix rubocop to count the template tests in coverage report

v3.0.0
------
### Improvement
- Improve configuration key naming
- Add common default attributes
  - WARNING: This change will set the HostKey attribute by default, possibly changing your server fingerprint.  See [Issue #5](https://github.com/rackspace-cookbooks/rackspace_openssh/issues/5)

v2.0.0
------
### Improvement
Rackspace rework

v1.3.0
------
### Improvement
- **[COOK-3644](https://tickets.opscode.com/browse/COOK-3644)** - Add FreeBSD support
- **[COOK-2517](https://tickets.opscode.com/browse/COOK-2517)** - Add hash support
- **[COOK-2000](https://tickets.opscode.com/browse/COOK-2000)** - Make mode of sshd_config a configurable option

### Bug
- **[COOK-3558](https://tickets.opscode.com/browse/COOK-3558)** - Fix RSA Server Config Options
- **[COOK-3557](https://tickets.opscode.com/browse/COOK-3557)** - Fix PubkeyAuthentication option


v1.2.2
------
### Bug
- **[COOK-3304](https://tickets.opscode.com/browse/COOK-3304)** - Fix error setting Dynamic `ListenAddresses`

v1.2.0
------
### Improvement
- [COOK-2647]: `port_ssh` iptables template has no corresponding recipe

v1.1.4
------
- [COOK-2225] - Add platform_family suse

v1.1.2
------
- [COOK-1954] - Fix attribute camel case to match `man sshd_config`
- [COOK-1889] - SSH restarting on each chef run due to template changes

v1.1.0
------
- [COOK-1663] - Configurable ListenAddress based off list of interface names
- [COOK-1685] - Make default sshd_config value more robust

v1.0.0
------
- [COOK-1014] - Templates for ssh(d).conf files.

v0.8.1
------
- Current public release


spec/spec_helper.rb
require 'berkshelf'
require 'chefspec'

spec/unit/recipes/default_spec.rb
require 'spec_helper'

describe 'rackspace_openssh::default' do
  let(:chef_run) { ChefSpec::Runner.new.converge(described_recipe) }

  it 'installs the openssh packages' do
    expect(chef_run).to install_package('openssh-client')
    expect(chef_run).to install_package('openssh-server')
  end

  it 'starts the ssh service' do
    expect(chef_run).to start_service('ssh')
    expect(chef_run).to enable_service('ssh')
  end

  it 'writes the ssh_config' do
    expect(chef_run).to create_template('/etc/ssh/ssh_config').with(
      user: 'root',
      group: 'root',
      mode: '0644'
    )
  end

  it 'writes the ssh_config' do
    expect(chef_run).to create_template('/etc/ssh/sshd_config').with(
      user: 'root',
      group: 'root',
      mode: '0644'
    )
  end

  it 'has expected sshd_config settings' do
    expect(chef_run).to render_file('/etc/ssh/sshd_config').with_content('LogLevel VERBOSE')
  end

  it 'does not allow root logins' do
    expect(chef_run).to render_file('/etc/ssh/sshd_config').with_content('PermitRootLogin no')
  end
end


test/integration/default/serverspec/default_spec.rb
require 'spec_helper'

describe 'rackspace_openssh::default' do
  it 'starts the ssh service' do
    case backend(Serverspec::Commands::Base).check_os
    when 'Debian'
      expect(service('ssh')).to be_enabled
      expect(service('ssh')).to be_running
    when 'RedHat'
      expect(service('sshd')).to be_enabled
      expect(service('sshd')).to be_running
    end
  end

  it 'is listening on the configured port' do
    expect(port(22)).to be_listening
  end
end


attributes/default.rb
#
# Cookbook Name:: rackspace_openssh
# Attributes:: default
#
# Author:: Ernie Brodeur <ebrodeur@ujami.net>
# Copyright 2008-2013, Opscode, Inc.
# Copyright 2014, Rackspace, US Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# Attributes are commented out using the default config file values.
# Uncomment the ones you need, or set attributes in a role.
#

default['rackspace_openssh']['templates_cookbook']['ssh_config'] = 'rackspace_openssh'
default['rackspace_openssh']['templates_cookbook']['sshd_config'] = 'rackspace_openssh'

default['rackspace_openssh']['config']['package_name'] = case node['platform_family']
                                                         when 'rhel'
                                                           %w(openssh-clients openssh)
                                                         else
                                                           %w(openssh-client openssh-server)
                                                         end

default['rackspace_openssh']['config']['service_name'] = case node['platform_family']
                                                         when 'rhel'
                                                           'sshd'
                                                         else
                                                           'ssh'
                                                         end

default['rackspace_openssh']['config']['config_mode'] = case node['platform_family']
                                                        when 'rhel'
                                                          '0600'
                                                        else
                                                          '0644'
                                                        end

default['rackspace_openssh']['config']['rootgroup']    = 'root'

# ssh config group
default['rackspace_openssh']['config']['client']['host'] = '*'
# default['rackspace_openssh']['config']['client']['forward_agent'] = 'no'
# default['rackspace_openssh']['config']['client']['forward_x11'] = 'no'
# default['rackspace_openssh']['config']['client']['rhosts_rsa_authentication'] = 'no'
# default['rackspace_openssh']['config']['client']['rsa_authentication'] = 'yes'
# default['rackspace_openssh']['config']['client']['password_authentication'] = 'yes'
# default['rackspace_openssh']['config']['client']['host_based_authentication'] = 'no'
# default['rackspace_openssh']['config']['client']['gssapi_authentication'] = 'no'
# default['rackspace_openssh']['config']['client']['gssapi_delegate_credentials'] = 'no'
# default['rackspace_openssh']['config']['client']['batch_mode'] = 'no'
# default['rackspace_openssh']['config']['client']['check_host_ip'] = 'yes'
# default['rackspace_openssh']['config']['client']['address_family'] = 'any'
# default['rackspace_openssh']['config']['client']['connect_timeout'] = '0'
# default['rackspace_openssh']['config']['client']['strict_host_key_checking'] = 'ask'
# default['rackspace_openssh']['config']['client']['identity_file'] = '~/.ssh/identity'
# default['rackspace_openssh']['config']['client']['identity_file_rsa'] = '~/.ssh/id_rsa'
# default['rackspace_openssh']['config']['client']['identity_file_dsa'] = '~/.ssh/id_dsa'
# default['rackspace_openssh']['config']['client']['port'] = '22'
# default['rackspace_openssh']['config']['client']['protocol'] = [ '2 1' ']
# default['rackspace_openssh']['config']['client']['cipher'] = '3des'
# default['rackspace_openssh']['config']['client']['ciphers'] = [ 'aes128-ctr aes192-ctr aes256-ctr arcfour256 arcfour128 aes128-cbc 3des-cbc' ']
# default['rackspace_openssh']['config']['client']['macs'] = [ 'hmac-md5 hmac-sha1 umac-64@openssh.com hmac-ripemd160' ']
# default['rackspace_openssh']['config']['client']['escape_char'] = '~'
# default['rackspace_openssh']['config']['client']['tunnel'] = 'no'
# default['rackspace_openssh']['config']['client']['tunnel_device'] = 'any:any'
# default['rackspace_openssh']['config']['client']['permit_local_command'] = 'no'
# default['rackspace_openssh']['config']['client']['visual_host_key'] = 'no'
# default['rackspace_openssh']['config']['client']['proxy_command'] = 'ssh -q -W %h:%p gateway.example.com'

# sshd config group
# set verbose logging above default precedence
normal['rackspace_openssh']['config']['server']['LogLevel'] = 'VERBOSE'

default['rackspace_openssh']['config']['server']['Port'] = '22'
default['rackspace_openssh']['config']['server']['Protocol'] = '2'
default['rackspace_openssh']['config']['server']['HostKey'] = '/etc/ssh/ssh_host_rsa_key'
default['rackspace_openssh']['config']['server']['ChallengeResponseAuthentication'] = 'no'
default['rackspace_openssh']['config']['server']['KeyRegenerationInterval'] = '1h'
default['rackspace_openssh']['config']['server']['ServerKeyBits'] = '1024'
default['rackspace_openssh']['config']['server']['SyslogFacility'] = 'AUTH'
default['rackspace_openssh']['config']['server']['LoginGraceTime'] = '2m'
default['rackspace_openssh']['config']['server']['PermitRootLogin'] = 'no'
default['rackspace_openssh']['config']['server']['StrictModes'] = 'yes'
default['rackspace_openssh']['config']['server']['MaxAuthTries'] = '6'
default['rackspace_openssh']['config']['server']['RSAAuthentication'] = 'yes'
default['rackspace_openssh']['config']['server']['PubkeyAuthentication'] = 'yes'
default['rackspace_openssh']['config']['server']['AuthorizedKeysFile'] = '%h/.ssh/authorized_keys'
default['rackspace_openssh']['config']['server']['IgnoreRhosts'] = 'yes'
default['rackspace_openssh']['config']['server']['HostBasedAuthentication'] = 'no'
default['rackspace_openssh']['config']['server']['PermitEmptyPasswords'] = 'no'
default['rackspace_openssh']['config']['server']['PasswordAuthentication'] = 'yes'
default['rackspace_openssh']['config']['server']['ChallengeResponseAuthentication'] = 'no'
default['rackspace_openssh']['config']['server']['GSSAPICleanUpCredentials'] = 'yes'
default['rackspace_openssh']['config']['server']['X11Forwarding'] = 'no'
default['rackspace_openssh']['config']['server']['PrintLastlog'] = 'yes'
default['rackspace_openssh']['config']['server']['TCPKeepAlive'] = 'yes'
default['rackspace_openssh']['config']['server']['AcceptEnv'] = 'LANG LC_* XMODIFIERS
'
default['rackspace_openssh']['config']['server']['Subsystem'] = case node['platform_family']
                                                                when 'rhel'
                                                                  'sftp /usr/libexec/sftp-server'
                                                                else
                                                                  'sftp /usr/lib/openssh/sftp-server'
                                                                end
default['rackspace_openssh']['config']['server']['UsePAM'] = 'yes'

# Other common sshd options
# default['rackspace_openssh']['config']['server']['HostKey'] = '/etc/ssh/ssh_host_dsa_key'
# default['rackspace_openssh']['config']['server']['HostKey'] = '/etc/ssh/ssh_host_ecdsa_key'
# default['rackspace_openssh']['config']['server']['AddressFamily'] = 'any'
# default['rackspace_openssh']['config']['server']['ListenAddress'] = [ '0.0.0.0 ::' ']
# default['rackspace_openssh']['config']['server']['HostKeyV1'] = '/etc/ssh/ssh_host_key'
# default['rackspace_openssh']['config']['server']['MaxSessions'] = '10'
# default['rackspace_openssh']['config']['server']['RhostsRSAAuthentication'] = 'no'
# default['rackspace_openssh']['config']['server']['IgnoreUserKnownHosts'] = 'no'
# default['rackspace_openssh']['config']['server']['KerberosAuthentication'] = 'no'
# default['rackspace_openssh']['config']['server']['KerberosOrLocalpasswd'] = 'yes'
# default['rackspace_openssh']['config']['server']['KerberosTicketCleanup'] = 'yes'
# default['rackspace_openssh']['config']['server']['KerberosGetAfsToken'] = 'no'
# default['rackspace_openssh']['config']['server']['GSSAPIAuthentication'] = 'no'
# default['rackspace_openssh']['config']['server']['AllowAgent_forwarding'] = 'yes'
# default['rackspace_openssh']['config']['server']['AllowTcpForwarding'] = 'yes'
# default['rackspace_openssh']['config']['server']['GatewayPorts'] = 'no'
# default['rackspace_openssh']['config']['server']['X11DisplayOffset'] = '10'
# default['rackspace_openssh']['config']['server']['X11UseLocalhost'] = 'yes'
# default['rackspace_openssh']['config']['server']['PrintMotd'] = 'yes'
# default['rackspace_openssh']['config']['server']['UseLogin'] = 'no'
# default['rackspace_openssh']['config']['server']['UsePrivilegeSeparation'] = 'yes'
# default['rackspace_openssh']['config']['server']['PermitUserEnvironment'] = 'no'
# default['rackspace_openssh']['config']['server']['Compression'] = 'delayed'
# default['rackspace_openssh']['config']['server']['ClientAliveInterval'] = '0'
# default['rackspace_openssh']['config']['server']['ClientAliveCountMax'] = '3'
# default['rackspace_openssh']['config']['server']['UseDns'] = 'yes'
# default['rackspace_openssh']['config']['server']['PidFile'] = '/var/run/sshd.pid'
# default['rackspace_openssh']['config']['server']['MaxStartups'] = '10'
# default['rackspace_openssh']['config']['server']['PermitTunnel'] = 'no'
# default['rackspace_openssh']['config']['server']['ChrootDirectory'] = 'none'
# default['rackspace_openssh']['config']['server']['Banner'] = 'none'
